### PR TITLE
[ CEPH-ADM ]: Restructured Cephadm modules and test-suites

### PR DESCRIPTION
**Pull request by sunilkumarn417**
- Restructured cephadm modules to more modular way.
- Added separate test-suites for deployment and cephadm sanity.
- Code changes to collect failure events in case of service/daemon deployment failure.

**Test Results:**
- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611802119142/ ( whole cluster deployment).
- http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611803879675/ ( Cephadm sanity - currently host operations included).

### CODE REVIEW

File: ceph/ceph_admin/__init__.py [hunk 1]
```diff
@@ -0,0 +1,750 @@
...
+        err = err.read().decode().strip()
+        logger.debug(out)
+        return out, err
+
+    def set_tool_repo(self):
+        """
+        Enable Ceph tool repo on every cluster node
+        """
+        for node in self.cluster.get_nodes():
+            node.exec_command(
```
Comment (psathyan) [inline ceph/ceph_admin/__init__.py:174]: IMO, `node.exec_command(sudo=True, cmd="yum-config-manager --add...")`. Let's keep it uniform...

File: ceph/ceph_admin/__init__.py [hunk 2]
```diff
@@ -0,0 +1,750 @@
...
+        """
+        logger.info("Installing cephadm")
+        self.installer.exec_command(
+            sudo=True,
+            cmd="yum install cephadm -y --nogpgcheck",
+            long_running=True,
+        )
+
+        if kw.get("upgrade"):
+            self.installer.exec_command(cmd="sudo yum update metadata")
```
Comment (psathyan) [inline ceph/ceph_admin/__init__.py:198]: Opt to set `sudo` as an argument to `exec_command`

File: ceph/ceph_admin/bootstrap_mixin.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        self.cluster.setup_ssh_keys()
+
+        # set tool download repository
+        self.set_tool_repo()
+
+        # install/download cephadm package on installer
+        self.install_cephadm()
+
+        # Create and set permission to ceph directory
+        self.installer.exec_command(cmd="sudo mkdir -p /etc/ceph")
```
Comment (psathyan) [inline ceph/ceph_admin/bootstrap_mixin.py:31]: Let's opt to use `sudo=True` as an argument to `exec_command`

File: ceph/ceph_admin/host_mixin.py [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+        """
+        Fetch labels attach to a node by hostname
+        Returns:
+            labels: list of labels
+        """
+        nodes = self.host_list()
+        for node in nodes:
+            if node_name in node["hostname"]:
+                return node["labels"]
+        return []
```
Comment (psathyan) [inline ceph/ceph_admin/host_mixin.py:234]: Would it be a good idea to raise an exception `raise ResourceNotFound`?

File: ceph/ceph_admin/host_mixin.py [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+        """
+        Returns ip address of attached node using hostname
+
+        Returns:
+            ip_address: IP address of host name
+        """
+        for node in self.host_list():
+            if node_name in node["hostname"]:
+                return node["addr"]
+        return None
```
Comment (psathyan) [inline ceph/ceph_admin/host_mixin.py:262]: IMO, we could raise an exception if the resource is not found.

File: conf/pacific/cephadm/sanity-cephadm.yaml [hunk 6]
```diff
@@ -1,51 +1,51 @@
...
+      node5:
+        role:
+          - osd
+          - node-exporter
+        no-of-volumes: 4
+        disk-size: 15
+      node6:
+        role:
           - mon
           - rgw
```
Comment (psathyan) [inline conf/pacific/cephadm/sanity-cephadm.yaml:40]: As there are two `rgw` nodes, do you suggest we try load balancing or HAproxy?

**Review by psathyan — COMMENTED** [review]
I like the way you have documented the modules and methods. Appreciate the good work.

There are some minor comments and queries.

File: ceph/ceph_admin/host_mixin.py (hunk 4 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/host_mixin.py:236]: Currently used to validate label in another method, would keep it to next improvements.

File: ceph/ceph_admin/host_mixin.py (hunk 5 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/host_mixin.py:264]: Currently used to validate label in another method, would keep it to next improvements.

File: conf/pacific/cephadm/sanity-cephadm.yaml (hunk 6 above)
Comment (sunilkumarn417) [reply conf/pacific/cephadm/sanity-cephadm.yaml:40]: Would consider this in next cephadm commits when service/daemons operations are handled.

File: ceph/ceph_admin/bootstrap_mixin.py (hunk 3 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/bootstrap_mixin.py]: Done

File: ceph/ceph_admin/__init__.py (hunk 2 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/__init__.py]: Done

File: ceph/ceph_admin/__init__.py [hunk 7]
```diff
@@ -0,0 +1,750 @@
...
+            for _id in ids:
+                if _id in daemons:
+                    if daemons[_id] == "running":
+                        count += 1
+
+            logger.info("%s/%s %s daemon(s) up...." % (count, len(ids), daemon))
+            if count == len(ids):
+                return True
+
+            logger.info("re-checking... %s times left", checks)
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply ceph/ceph_admin/__init__.py]: Combined both logs into one now.

File: ceph/ceph_admin/__init__.py [hunk 8]
```diff
@@ -0,0 +1,750 @@
...
+
+        # Checkout service events, in case of processes not-running
+        out, _ = self.shell(
+            remote=self.installer,
+            args=["ceph", "orch", "ls", "-f", "json-pretty"],
+        )
+
+        for i in json.loads(out):
+            if daemon in i.get("service_type") and i["status"].get("running") == 0:
+                logger.exception("Service status(es) : %s" % i)
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply ceph/ceph_admin/__init__.py]: Changed to logger.error level.

File: ceph/ceph_admin/__init__.py [hunk 9]
```diff
@@ -0,0 +1,750 @@
+"""
+[ CEPHADM ] Ceph Administrator module enables ceph pre-requisites
```
[the remark this answers is not in the document]
Comment (sunilkumarn417) [reply ceph/ceph_admin/__init__.py]: done

**Comment by sunilkumarn417** [comment]
**Test results:** 

http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611854702539/ (deployment)
http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-1611856510307/ (Cephadm sanity)

File: ceph/ceph_admin/__init__.py [hunk 10]
```diff
@@ -0,0 +1,764 @@
...
+            for _id in ids:
+                if _id in daemons:
+                    if daemons[_id] == "running":
+                        count += 1
+
+            logger.info(
+                "%s/%s %s daemon(s) up... Retries: %s"
+                % (count, len(ids), daemon, checks)
+            )
+            if count == len(ids):
```
Comment (psathyan) [inline ceph/ceph_admin/__init__.py:134]: Would it make sense moving this statement above the log statement?

**Review by psathyan — COMMENTED** [review]
Looks good to me... one minor query

File: ceph/ceph_admin/__init__.py (hunk 10 above)
Comment (sunilkumarn417) [reply ceph/ceph_admin/__init__.py:134]: Yes we can, but we may need log statement to say "all daemons are running" before returning True value, that means progressively we may miss "2/2 daemons up and running" before returning.

or may be we can add logger statement to log "all daemons are up and running" before returning.

I also think that currently both(progress and final statuses) dumped in current logging with single log expression by not moving after check `count == len(ids)`.

Comment (psathyan) [reply ceph/ceph_admin/__init__.py:134]: Okay, I agree with your assessment.

**Review by psathyan — APPROVED** [review]
Looks good to me

**Pull request merged**